internal/launcher/launcher.go

```diff
@@ -0,0 +1,158 @@
+		run:       runK6Cmd,
+	}
+}
+
+// launch executes k6 either by launching a provisioned binary or defaulting to the
+// current binary it this is not necessary.
+// If the fhe fallback is called, it can exit the process so don't assume it will return
+func (l *launcher) launch() {
+	// if binary provisioning not enabled, continue with regular k6 execution path
+	if !l.gs.Flags.BinaryProvisioning {
+		l.gs.Logger.Debug("binary provisioning disabled")
+		l.fallback(l.gs)
+		return
+	}
+
+	l.gs.Logger.Info("trying to provision binary")
+
+	deps, err := k6deps.Analyze(newDepsOptions(l.gs, l.gs.CmdArgs[1:]))
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("failed to analyze dependencies, can't try binary provisioning, please report this issue")
+		l.gs.OSExit(1)
+	}
+
+	// binary provisioning enabled but not required by this command
+	// continue with regular k6 execution path
+	if !isCustomBuildRequired(build.Version, deps) {
+		l.gs.Logger.
+			Debug("binary provisioning not required")
+		l.fallback(l.gs)
+		return
+	}
+
+	l.gs.Logger.
+		WithField("deps", deps).
+		Info("dependencies identified, binary provisioning required")
+
+	l.launchCustomBuild(deps)
+}
+
+func (l *launcher) launchCustomBuild(deps k6deps.Dependencies) {
+	// get the k6 binary from the build service
+	binPath, versions, err := l.provision(l.gs, deps)
+	if err != nil {
+		l.gs.Logger.
+			WithError(err).
+			Error("failed to fetch a binary with required dependencies, please report this issue")
+		l.gs.OSExit(1)
+		// in tests calling l.gs.OSExit does not ends execution so we have to return
+		return
+	}
+
+	l.gs.Logger.
+		Info("k6 has been provisioned with version(s) ", versions)
+
+	l.gs.Logger.Debug("launching provisioned k6 binary")
+
+	// execute provisioned binary
+	if rc, err := l.run(l.gs, binPath); err != nil {
+		l.gs.Logger.Error(err)
+		l.gs.OSExit(rc)
+	}
+}
+
+// runK6Cmd runs the k6 binary passing the original arguments
+func runK6Cmd(gs *state.GlobalState, binPath string) (int, error) {
+	cmd := exec.CommandContext(gs.Ctx, binPath, gs.CmdArgs[1:]...) //nolint:gosec
+	cmd.Stderr = gs.Stderr
+	cmd.Stdout = gs.Stdout
+	cmd.Stdin = gs.Stdin
+
+	// disable binary provisioning to avoid a provisioning loop
+	gs.Env["K6_BINARY_PROVISIONING"] = "false"
+
+	if err := cmd.Run(); err != nil {
+		var eerr *exec.ExitError
+		if errors.As(err, &eerr) {
+			return eerr.ExitCode(), err
+		}
+	}
+
+	return 0, nil
+}
+
+// isCustomBuildRequired checks if the build is required
+// it's required if there is one or more dependencies other than k6 itself
+// or if the required k6 version is not satisfied by the current binary's version
+// TODO: get the version of any built-in extension and check if they satisfy the dependencies
+func isCustomBuildRequired(baseK6Version string, deps k6deps.Dependencies) bool {
+	if len(deps) == 0 {
+		return false
+	}
+
+	// Early return if there are multiple dependencies
+	if len(deps) > 1 {
+		return true
+	}
```

Comment:
I might be missing some context (around `k6deps`), but what's considered a dependency here? How do you know what extensions are available from the current binary?

Let's say this code is being executed by a binary compiled with `xk6` with support for `xk6-my-custom-extension`. Will it have `len(deps) > 1` anyway? 🤔

Reply:
If that's what's stated by the `TODO` in the code comment at the very beginning (together with the function definition), then I think this is likely a must before actually being ready to merge this PR? 🤔 

Otherwise, we may require a custom build for dozens of cases where it's not really needed. And I think this is quite critical because custom builds require network -downloading the new binary-, setting up cloud token, etc. Not just a bit of extra work.

Reply:
Even when I agree we should prevent this unnecessary binary provisioning (hence the TODO). However, it is important to consider that the binary provisioning is not enabled by default. If someone builds a custom binary with extensions, they have to enable the binary provisioning, what looks more like a misconfiguration.

Reply:
>If someone builds a custom binary with extensions, they have to enable the binary provisioning for this to happen, which looks more like a misconfiguration.

@pablochacin extensions are modules but also outputs. If I interpret the code right, we aren't addressing any case for extension outputs. So there is still a valid case for this to happen.

If I need both module and output extensions, does it mean it won't work?

Reply:
There's a way using pragmas as explained in `k6exec` documentation:

```
The pragma syntax can also be used to specify an extension dependency that is not referenced in an import expression. A typical example of this is the Output type extension such as [xk6-top](https://github.com/szkiba/xk6-top):

"use k6 with top >= 0.1";
```

We could also scan the CLI for `--out` flags, but there's no way to provide information about the version required for the extension.

In any case, at this moment the extension catalog used by the binary provisioning does not include output extensions. I suggest we handle this case in a follow-up PR.

Reply:
https://github.com/grafana/k6/issues/4720